src/tools/rbd/action/Ubbd.cc

```diff
@@ -0,0 +1,190 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/errno.h"
+#include "tools/rbd/Shell.h"
+#include "tools/rbd/Utils.h"
+#include <iostream>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/scope_exit.hpp>
+
+#if defined(WITH_UBBD)
+extern "C" {
+#include <libubbd.h>
+}
+#endif
+
+namespace rbd {
+namespace action {
+namespace ubbd {
+
+namespace at = argument_types;
+namespace po = boost::program_options;
+
+int execute_list(const po::variables_map &vm,
+                 const std::vector<std::string> &ceph_global_init_args) {
+#if defined(WITH_UBBD)
+  struct ubbdd_mgmt_rsp rsp = { 0 };
+  struct ubbd_list_options list_opts = { .type = UBBD_DEV_TYPE_RBD };
+  int r;
+  int i;
+
+  r = ubbd_list(&list_opts, &rsp);
+  if (r)
+    return r;
+
+  for (i = 0; i < rsp.list.dev_num; i++) {
+    std::cerr << "/dev/ubbd" << rsp.list.dev_list[i] << std::endl;
+  }
+
+  return 0;
+#else
+  std::cerr << "rbd: ubbd device is not supported" << std::endl;
+  return -EOPNOTSUPP;
+#endif
+}
+
+int execute_map(const po::variables_map &vm,
+                const std::vector<std::string> &ceph_global_init_args) {
+#if defined(WITH_UBBD)
+  size_t arg_index = 0;
+  std::string pool_name;
+  std::string nspace_name;
+  std::string image_name;
+  std::string snap_name;
+  struct ubbdd_mgmt_rsp rsp = { 0 };
+  struct ubbd_map_options opts = { 0 };
+  std::string dev_type = "rbd";
+
+  int r = utils::get_pool_image_snapshot_names(
+    vm, at::ARGUMENT_MODIFIER_NONE, &arg_index, &pool_name, &nspace_name,
+    &image_name, &snap_name, true, utils::SNAPSHOT_PRESENCE_PERMITTED,
+    utils::SPEC_VALIDATION_NONE);
+  if (r < 0) {
+    return r;
+  }
+  opts.type = opts.generic_dev.opts.type = dev_type.c_str();
+  opts.generic_dev.opts.rbd.pool = pool_name.c_str();
+  opts.generic_dev.opts.rbd.ns = nspace_name.c_str();
+  opts.generic_dev.opts.rbd.image = image_name.c_str();
+  opts.generic_dev.opts.rbd.snap = snap_name.c_str();
+
+  if (vm["read-only"].as<bool>()) {
+    opts.read_only = true;
+  }
+  if (vm["exclusive"].as<bool>()) {
+    std::cerr << "rbd: warning: exclusive is not supported" << std::endl;
+  }
+  if (vm["quiesce"].as<bool>()) {
+    std::cerr << "rbd: warning: quiesce is not supported" << std::endl;
+  }
+  if (vm.count("quiesce-hook")) {
+    std::cerr << "rbd: warning: quiesce-hook is not supported" << std::endl;
```

Comment:
If the primary goal to enable all librbd features, why are these options not supported?

Reply:
This PR is plan to support full *image features*, so quiesce snapshot is not supported. But the quiesce is not hard to support via do it in ubbd-backend process in later version.